core/src/main/scala/chisel3/Data.scala

```diff
@@ -372,6 +372,89 @@ trait BaseType extends HasId with NamedComponent {
     */
   private[chisel3] def cloneTypeFull: this.type
 
+  // Internal API: returns a ref that can be assigned to, if consistent with the binding
+  private[chisel3] def lref: Node = {
+    requireIsHardware(this)
+    requireVisible()
+    topBindingOpt match {
+      case Some(binding: ReadOnlyBinding) =>
+        throwException(s"internal error: attempted to generate LHS ref to ReadOnlyBinding $binding")
+      case Some(ViewBinding(target)) => reify(target).lref
+      case Some(binding: TopBinding) => Node(this)
+      case opt => throwException(s"internal error: unknown binding $opt in generating LHS ref")
+    }
+  }
+
+  // Internal API: returns a ref, if bound
+  private[chisel3] final def ref: Arg = {
+    def materializeWire(makeConst: Boolean = false): Arg = {
+      if (!Builder.currentModule.isDefined) throwException(s"internal error: cannot materialize ref for $this")
+      implicit val sourceInfo = UnlocatableSourceInfo
+      this match {
+        case (data: Data) => {
+          if (makeConst) {
+            WireDefault(Const(chiselTypeOf(data)), data).ref
+          } else {
+            WireDefault(data).ref
+          }
+        }
+        case _ => throwException(s"internal error: cannot materialize ref for $this")
+      }
+    }
+    requireIsHardware(this)
+    topBindingOpt match {
+      // DataView
+      case Some(ViewBinding(target)) => reify(target).ref
+      case Some(AggregateViewBinding(viewMap)) =>
+        viewMap.get(this.asInstanceOf[Data]) match { // AggregateViewBinding should only be possible on Data.
```

Comment:
I don't love having to have these casts as it suggests that this is a bit of false re-use here. I'm okay with it for now but we should think about how to refactor this as we implement Properties further.